data_preparation/predictor_data_prep.r

```diff
@@ -241,93 +247,146 @@ process_ts_raster <- function(region, ind_filename, indicator, crop_mask_filenam
-                  crop_mask_filename=crop_mask_filename)
+  dfs <- pblapply(file_list, process_ts_raster, indicator=indicator,
+                  crop_mask_file=crop_mask_file,
+                  region_boundaries=region_boundaries)
 
   result <- rbindlist(dfs)
+  return(result)
 }
 
-
 process_indicators <- function(crop, region,
                                sel_indicators,
                                start_year, end_year,
                                crop_mask_file, num_cpus) {
-  # print(region)
-
-  ###############################
-  # Process each indicator      #
-  ###############################
   for (indicator in sel_indicators) {
     filename_pattern <- indicators[[indicator]][["filename_pattern"]]
     indicator_source <- indicators[[indicator]][["source"]]
     is_time_series <- indicators[[indicator]][["is_time_series"]]
     is_categorical <- indicators[[indicator]][["is_categorical"]]
 
     # print(paste(indicator, indicator_source, filename_pattern, is_time_series, is_categorical))
+    region_boundaries <- get_shapes(region)
     crop_mask <- rast(crop_mask_file)
     crop_mask[crop_mask > 100] <- NA
     crop_mask[crop_mask < 0] <- NA
     result <- NULL
 
-    ###############
-    # static data #
-    ###############
-    if (!is_time_series) {
+    if (is_time_series) {
+      ####################
+      # Time series data #
+      ####################
+
+      # NOTE: AgERA5 data is split by region.
+      # So include region-specific directory in path.
+      if (indicator_source == "AgERA5") {
+        indicator_path = file.path(PREDICTORS_DATA_PATH, indicator_source,
+                                   paste(region, indicator_source, sep="_"), indicator)
+      } else {
+        indicator_path = file.path(PREDICTORS_DATA_PATH, indicator_source, indicator)
+      }
+
+      found_raster_file <- FALSE
+      while(!found_raster_file) {
+        file_list <- list.files(path=indicator_path,
+                                pattern=glob2rx(paste0(filename_pattern, as.character(start_year), "*")),
+                                full.names=TRUE)
+        if (length(file_list) == 0) {
+          warning(paste("Files not found for", start_year))
+          start_year <- start_year + 1 
+        } else {
+          found_raster_file <- TRUE
+        }
+      }
+
+      # Resample crop mask to indicator rastor resolution.
+      ind_rast <- rast(file_list[[1]])
+      crop_mask <- resample(crop_mask, ind_rast, method="bilinear")
+      crop_mask <- crop(crop_mask, region_boundaries)
+      # NOTE: We are overwriting in every iteration (per indicator).
+      # This is fine because resampling is needed per indicator.
+      resampled_filename <- file.path(AGML_ROOT, "crop_masks",
+                                      "resampled_crop_mask.tif")
+      writeRaster(crop_mask, resampled_filename, overwrite=TRUE)
+
+      rm(list=c("ind_rast", "crop_mask"))
```

Comment:
It could be useful to put `gc()` here too, as otherwise the newly spawned processes may still carry on the garbage (processes are just forked as-is in memory).